### DISCUSSION TITLE
Voltage Compensator Design

### DISCUSSION THREAD

**Discussion by jinningwang** [CURENT/andes#199] [Modeling] [answered]
Hi Hantao,

I want to develop a voltage compensator model ``IEEEVC``, which should be in a new group ``VoltComp``. The ``VoltComp`` will have ``vcomp`` as common variables.

The basic idea is that:
Given a generator (``SynGen``) idx, if:
1) there is no exciter on the generator, the ``VoltComp`` output ``vcomp`` will control the ``SynGen`` ``vf``.
2) there is an exciter on the generator, the ``VoltComp`` output ``vcomp`` will control the ``Exciter`` ``vin``. Here, I mean to modify the ``vin.e_str`` by an ``ExtAlgeb``, replacing the ``v`` with ``vcomp``.

Then it raises a question that, how to retrieve the exciter idx according to the generator idx?

Or do you have any suggestions about such a ``VoltComp`` design?

Regards,
Jinning

**Comment by cuihantao** [comment, marked as answer]
There are two issues:
- Is `VoltComp` supposed to control the generator's `vf` if the exciter is absent? Or say, is `VoltComp` only allowed when exciter is present?
- the second one should work! More precisely, it is to introduce an `ExtAlgeb` in `VoltComp` that modifies the `vin` of exciters, which is currently assumed to be `v`. 

Please figure out the first issue by checking references or trying out existing tools.

**Reply by jinningwang** [reply]
1. A ``VoltComp`` is allowed regardless of the presence of exciter. Proof: In the PSS/e modeling book's description of ``CCOMP4U1``, a ``CCOMP4U1`` is attached to a generator, and it retrieves the stator currents to compensate the voltage.
2. A question of implementation about (2): in a ``VoltComp``, we can have the ``syn`` as the generator idx. However, how to retrieve the exciter idx which are attached to the generator? Absolutely we can require users to input exciter idx numbers, but it will be redundant and not elegant.

**Reply by cuihantao** [reply]
For the second question, for the time being, let's ask the user to provide it. If interested, you can tinker with it using `self.Exciter = BackRef()` in generators. But that `BackRef` will be collected in the `System.setup()` phase and probably after `ExtVar` address retrival. In other words, the exciter indices may not be ready when accessing the exciter variables. This needs to be confirmed as I couldn't recall it. 

For the first one, we need to determine whether to access generator's `vf` or exciter's `vin` based on if an exciter exists.

**Reply by jinningwang** [reply]
Thanks for your reply!

The enhancement can be put as a further enhancement, so we can have a ready-to-use version for the time being.
I checked the excitation chapter of Dr. Chow's book and found the voltage compensator should be used with the exciter. Sorry for my previous misunderstanding. Then I should be able to implement it without trouble.

Regards,
Jinning

**Comment by cuihantao** [comment]
That will make or a lot easier. It does not make much sense if someone wants to compensate for voltages without exciter. The first thing to do is to install an exciter. Lol.

Regards,
Hantao Cui
